pkg/downwardmetrics/virtio-serial/server.go

```diff
@@ -0,0 +1,244 @@
+		_ = watcher.Close()
+	}(watcher)
+
+	err = watcher.Add(filepath.Dir(socketPath))
+	if err != nil {
+		return err
+	}
+
+	for {
+		exists, err := diskutils.FileExists(socketPath)
+		if err != nil {
+			return err
+		}
+
+		if exists {
+			break
+		}
+
+		select {
+		case <-watcher.Events:
+		case err, ok := <-watcher.Errors:
+			if !ok {
+				return errors.New("unknown error")
+			}
+			return err
+		}
+	}
+	return nil
+}
+
+func connect(socketPath string, retries uint) (net.Conn, error) {
+	var conn net.Conn
+	var err error
+
+	multiplier := 1
+	for i := uint(0); i < retries+1; i++ {
+		conn, err = net.Dial("unix", socketPath)
+		if err == nil {
+			break
+		}
+
+		// It is only tried again in case not one is
+		// listening on the other end yet
+		if !errors.Is(err, syscall.ECONNREFUSED) {
+			break
+		}
+
+		if i == retries {
+			break // it was the last attempt
+		}
+
+		backoff := time.Duration(multiplier) * time.Second
+		time.Sleep(backoff)
+		multiplier *= 2
+	}
+
+	return conn, err
+}
+
+type downwardMetricsServer struct {
+	retries            uint // Number of retires after the first connection fails
+	virtioSerialSocket string
+	scraper            *metricsScraper.Scraper
+	libvirtSocket      string
+}
+
+func (s *downwardMetricsServer) handleRequest(conn net.Conn, requestLine string) error {
+	requestPath, err := parseRequestLine(requestLine)
+	if err != nil {
+		replyError(conn)
+		return err
+	}
+
+	if requestPath != "/metrics/XML" {
+		replyError(conn)
+		return errors.New("invalid request")
+	}
+
+	var xmlMetrics []byte
+	metrics, err := s.scraper.Report(s.libvirtSocket)
+	if err != nil {
+		xmlMetrics = emptyMetrics
+		log.Log.Reason(err).Warningf("failed to collect the metrics")
+	} else {
+		xmlMetrics, err = xml.MarshalIndent(metrics, "", "  ")
+		if err != nil {
+			return fmt.Errorf("failed to encode metrics: %v", err)
+		}
+	}
+
+	// `vm-dump-metrics` expects `\n\n` as a termination symbol
+	xmlMetrics = append(xmlMetrics, '\n', '\n')
+
+	err = reply(conn, xmlMetrics)
+	if err != nil {
+		return fmt.Errorf("failed to send the metrics: %v", err)
+	}
+	return nil
+}
+
+// DownwardMetrics Virtio Server
+// By defining a virtio-serial device, QEMU creates a unix domain socket on
+// the host that is connected to a serial port in the VM.
+//
+// This server is a bit special, since it serves the process, inside the VM,
+// requests by replying with the DownwardMetrics. But it connects to the
+// socket at the client end, since the unix socket is created by QEMU.
+func (s *downwardMetricsServer) start() error {
+	err := waitForSocket(s.virtioSerialSocket)
+	if err != nil {
+		return err
+	}
+
+	conn, err := connect(s.virtioSerialSocket, s.retries)
+	if err != nil {
+		return err
+	}
+	defer func(conn net.Conn) {
+		_ = conn.Close()
+	}(conn)
+
+	for {
+		request, err := waitForRequest(conn)
+		if err != nil {
+			if errors.Is(err, io.EOF) {
+				return nil
+			}
+			return err
+		}
+
+		go func(conn net.Conn, request string) {
```

Comment:
I don't think it is a good idea to hand out the `conn` to possibly multiple running go-routines. Can you instead return the results in a chan and have one routine the write the replies to the `conn` synchronously?

Reply:
Fixed